test/libethereum/Blockchain.cpp

```diff
@@ -0,0 +1,261 @@
+
+		RLPStream blockStream(3);
+		blockStream.appendRaw(rlpStream.out());
+		blockStream.appendRaw(txStream.out());
+		blockStream.appendRaw(uncleStream.out());
+
+		//store block to the blockchain
+		try
+		{
+			m_blockchain->sync(uncleBlockQueue, m_state.db(), 4);
+			m_blockchain->attemptImport(blockStream.out(), m_state.db());
+			m_state.sync(*m_blockchain.get());
+		}
+		catch (Exception const& _e)
+		{
+			cnote << "block sync or import did throw an exception: " << diagnostic_information(_e);
+			m_testBlocks.pop_back();
+			return;
+		}
+		catch (std::exception const& _e)
+		{
+			cnote << "block sync or import did throw an exception: " << _e.what();
+			m_testBlocks.pop_back();
+			return;
+		}
+	}
+
+	dev::eth::BlockChain& get()
+	{
+		return *m_blockchain.get();
+	}
+
+	void rebuild(WithExisting _withExisiting = WithExisting::Verify)
+	{
+		m_blockchain.reset();
+		m_blockchain.reset(new dev::eth::CanonBlockChain(m_td.path(), _withExisiting));
+	}	
+
+private:
+	dev::TransientDirectory m_td;				//BlockChain Dir
+	dev::TransientDirectory m_td2;				//State Dir
+	std::unique_ptr<dev::eth::BlockChain> m_blockchain;
+	dev::eth::State m_state;	
+	std::vector<TestBlock> m_testBlocks;
+};
+
+std::string const c_genesisInfoTest =
+R"ETHEREUM(
+{
+	"a94f5374fce5edbc8e2a8697c15331677e6ebf0b": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"0000000000000000000000000000000000000001": { "wei": "1" },
+	"0000000000000000000000000000000000000002": { "wei": "1" },
+	"0000000000000000000000000000000000000003": { "wei": "1" },
+	"0000000000000000000000000000000000000004": { "wei": "1" },
+	"dbdbdb2cbd23b783741e8d7fcf51e459b497e4a6": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"e6716f9544a56c530d868e4bfbacb172315bdead": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"b9c015918bdaba24b4ff057a92a3873d6eb201be": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"1a26338f0d905e295fccb71fa9ea849ffa12aaf4": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"2ef47100e0787b915105fd5e3f4ff6752079d5cb": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"cd2a3d9f938e13cd947ec05abc7fe734df8dd826": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"6c386a4b26f73c802f34673f7248bb118f97424a": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+	"e4157b34ea9615cfbde6b4fda419828124b70c78": { "wei": "1606938044258990275541962092341162602522202993782792835301376" },
+}
+)ETHEREUM";
+
+BOOST_AUTO_TEST_SUITE(libethereum)
+
+BOOST_AUTO_TEST_CASE(canonBlockChain)
+{
+	std::string genesisTemp = dev::eth::c_genesisInfo;
+
+	//A trick for blockchain to load changed genesis state
+	std::string *genesis_ptr;
+	genesis_ptr = const_cast<std::string*>( &dev::eth::c_genesisInfo );
+	*genesis_ptr = c_genesisInfoTest;
+
+	TestBlockChain blChain;
```

Comment:
use a tmp directory, otherwise you may corrupt data in the curent blockchain in .ethereum, or kill it (see line 215)